cmd/swarm/download.go

```diff
@@ -0,0 +1,144 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/cmd/utils"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/swarm/api"
+	swarm "github.com/ethereum/go-ethereum/swarm/api/client"
+	"gopkg.in/urfave/cli.v1"
+)
+
+func download(ctx *cli.Context) {
+	//isRecursive := false
+
+	log.Debug("swarm download")
+	args := ctx.Args()
+	if len(args) < 1 {
+		utils.Fatalf("Usage: swarm download <bzz locator> [<destination path>]")
+	}
+
+	var (
+		bzzapi      = strings.TrimRight(ctx.GlobalString(SwarmApiFlag.Name), "/")
+		isRecursive = ctx.GlobalBool(SwarmRecursiveUploadFlag.Name)
+		client      = swarm.NewClient(bzzapi)
+	)
+
+	dir := ""
+	filename := ""
+
+	if len(args) == 1 {
+		// no destination arg - assume current terminal working dir
+		workingDir, err := filepath.Abs("./")
+		log.Trace(fmt.Sprintf("swarm download: no destination path - assuming working dir: %s", workingDir))
+
+		if err != nil {
+			utils.Fatalf("Fatal: could not get current working directory")
+		}
+		dir = workingDir
+	} else {
+		log.Trace(fmt.Sprintf("swarm download: destination path arg: %s", args[1]))
+		dir = args[1]
+	}
+
+	log.Debug(fmt.Sprintf("working dir: %s", dir))
+
+	fi, err := os.Stat(dir)
+	if err != nil {
+		utils.Fatalf("could not stat path")
+	}
+
+	switch mode := fi.Mode(); {
+	case mode.IsRegular():
+		utils.Fatalf("destination path is not a directory!")
+	}
+
+	uri, err := api.Parse(args[0])
+
+	// assume behaviour according to --recursive switch
+	manifestList, err := client.List(uri.Addr, uri.Path)
+	if err != nil {
+		utils.Fatalf("could not list manifest: %v", err)
+	}
+
+	if isRecursive {
+		//we are downloading a directory, check that there's something there
+		if len(manifestList.CommonPrefixes)+len(manifestList.Entries) == 0 {
+			log.Info("the manifest address provided does not contain any entries to download")
+			os.Exit(0) //graceful shutdown
+		}
+		if err := client.DownloadDirectory(uri.Addr, uri.Path, dir); err != nil {
+			utils.Fatalf("encoutered a fatal error while downloading directory: %v", err)
+		}
+	} else {
+		// we are downloading a file
+		switch len(manifestList.Entries) {
+		case 0:
+			utils.Fatalf("could not find path requested at manifest address. make sure the path you've specified is correct")
+		case 1:
+			//continue
+		default:
+			utils.Fatalf("got too many matches for this path")
+		}
+
+		log.Debug(fmt.Sprintf("swarm download: downloading file/path from a manifest. hash: %s, path:%s", uri.Addr, uri.Path))
+
+		file, err := client.Download(uri.Addr, uri.Path)
+		if err != nil {
+			utils.Fatalf("could not download %s from given address: %s. error: %v", uri.Path, uri.Addr, err)
+		}
+
+		re := regexp.MustCompile("[^/]+$") //everything after last slash
```

Comment:
what is this logic after this line i dont get it.
If download is not handled by the client call, it should be fixed there no?